### PR TITLE
Use active ESC when only one can be displayed

### PR DESCRIPTION
**Pull request by andyp1per**
Currently the OSD displays ESC 0 only. This attempts to show the most relevant information

### CODE REVIEW

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 1]
```diff
@@ -2004,7 +2004,7 @@ void AP_OSD_Screen::draw_esc_temp(uint8_t x, uint8_t y)
 {
     int16_t etemp;
     // first parameter is index into array of ESC's.  Hardwire to zero (first) for now.
```
Comment (IamPete1) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:2006]: Comment is now wrong. I also think you should update the X Y params for this to explain its now the highest.

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp [hunk 2]
```diff
@@ -120,6 +120,30 @@ uint32_t AP_ESC_Telem::get_active_esc_mask() const {
...
+    uint32_t now_us = AP_HAL::micros();
+    for (uint8_t i = 0; i < ESC_TELEM_MAX_ESCS; i++) {
+        if (_telem_data[i].last_update_ms == 0 && !was_rpm_data_ever_reported(_rpm_data[i])) {
+            // have never seen telem from this ESC
+            continue;
+        }
+        if (_telem_data[i].stale(now)
+            && !rpm_data_within_timeout(_rpm_data[i], now_us, ESC_RPM_DATA_TIMEOUT_US)) {
+            continue;
+        }
```
Comment (IamPete1) [inline libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:138]: I think this is confusing. I would rather see that the ESC is not working than have it change to the next one and hide the issue.

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp [hunk 3]
```diff
@@ -120,6 +120,30 @@ uint32_t AP_ESC_Telem::get_active_esc_mask() const {
...
+        if (_rpm_data[i].rpm > max_rpm) {
+            max_rpm = _rpm_data[i].rpm;
+            ret = i;
+        }
```
Comment (IamPete1) [inline libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:142]: Its actually picking the one with the highest RPM?

**Review by IamPete1 — DISMISSED** [review]

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp (hunk 2 above)
Comment (IamPete1) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:138]: Maybe just two OSD items? "ESC max of all" which is what I think this code does? and add a new index param to the current ESC item with the default being to do whatever it is we do now. 

Another option might be to display the ESC number which the data is for alongside the data in the OSD.

File: libraries/AP_OSD/AP_OSD_Screen.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_OSD/AP_OSD_Screen.cpp]: Fixed

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:138]: Fixed

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:142]: Fixed

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 4]
```diff
@@ -1166,6 +1166,13 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info2[] = {
     AP_SUBGROUPINFO(rc_lq, "RC_LQ", 9, AP_OSD_Screen, AP_OSD_Setting),
 #endif
 
+#if HAL_WITH_ESC_TELEM
+    // @Param: ESC_IDX
+    // @DisplayName: ESC_IDX
+    // @Description: Index of the ESC to use for displaying ESC information. 0 means use the most active ESC.
```
Comment (IamPete1) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:1172]: I'm not a fan of "most active". What does it mean? Maybe just "highest RPM".

Comment (andyp1per) [reply libraries/AP_OSD/AP_OSD_Screen.cpp]: Fixed, plus the other bad bug you didn't spot ;)

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp (hunk 3 above)
Comment (IamPete1) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:142]: This is still picking the highest?

Comment (IamPete1) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:142]: Ah, I think the plan is now that that is the correct thing to do. Just the function name and comment don't say that.

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp [hunk 5]
```diff
@@ -120,6 +120,30 @@ uint32_t AP_ESC_Telem::get_active_esc_mask() const {
     return ret;
 }
 
+// return an active ESC for the purposes of reporting (e.g. in the OSD)
+uint8_t AP_ESC_Telem::get_active_esc() const
```
Comment (IamPete1) [inline libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:124]:
```suggestion
// return the index of the ESC with the highest RPM
uint8_t AP_ESC_Telem::get_max_RPM_index() const
```

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp [hunk 6]
```diff
@@ -120,6 +120,30 @@ uint32_t AP_ESC_Telem::get_active_esc_mask() const {
...
+{
+    uint32_t ret = 0;
+    float max_rpm = 0;
+    const uint32_t now = AP_HAL::millis();
+    uint32_t now_us = AP_HAL::micros();
```
Comment (peterbarker) [inline libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:129]:
```suggestion
    const uint32_t now_us = AP_HAL::micros();
```

**Comment by tridge** [comment]
can be merged once name is changed from "active" to something indicating highest RPM

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.cpp:142]: fixed

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp (hunk 5 above)
Comment (andyp1per) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.cpp]: fixed

File: libraries/AP_ESC_Telem/AP_ESC_Telem.cpp (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_ESC_Telem/AP_ESC_Telem.cpp]: fixed

**Review by tridge — APPROVED** [review]

**Pull request merged**